tests/tensorbox/test_tensorbox.py

```diff
@@ -0,0 +1,541 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for TensorBox subclasses"""
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane.tensorbox.numpy_box import NumpyBox
+
+
+class TestNumpyBox:
+    """Tests for the NumpyBox class"""
+
+    def test_creation_from_list(self):
+        """Test that a NumpyBox is automatically created from a list"""
+        x = [0.1, 0.2, 0.3]
+        res = qml.TensorBox(x)
+        assert isinstance(res, NumpyBox)
+        assert res.interface == "numpy"
+        assert isinstance(res.unbox(), np.ndarray)
+        assert np.all(res.unbox() == x)
+
+    def test_creation_from_tuple(self):
+        """Test that a NumpyBox is automatically created from a tuple"""
+        x = (0.1, 0.2, 0.3)
+        res = qml.TensorBox(x)
+        assert isinstance(res, NumpyBox)
+        assert res.interface == "numpy"
+        assert isinstance(res.unbox(), np.ndarray)
+        assert np.all(res.unbox() == x)
+
+    def test_creation_from_tensorbox(self):
+        """Test that a tensorbox input simply returns itself"""
+        x = qml.TensorBox(np.array([0.1, 0.2, 0.3]))
+        res = qml.TensorBox(x)
+        assert x is res
+
+    def test_unknown_input_type(self):
+        """Test that an exception is raised if the input type
+        is unknown"""
+        with pytest.raises(ValueError, match="Unknown tensor type"):
+            qml.TensorBox(True)
+
+    def test_len(self):
+        """Test length"""
+        x = np.array([[1, 2], [3, 4]])
+        res = qml.TensorBox(x)
+        assert len(res) == len(x) == 2
+
+    def test_ufunc_compatibility(self):
+        """Test that the NumpyBox class has ufunc compatibility"""
+        x = np.array([0.1, 0.2, 0.3])
+        res = np.sum(np.sin(qml.TensorBox(x)))

```

Comment:
Oh ufunc is for vectorised evalaution?

Reply:
Yep, `ufunc` is a NumPy-aware function that automatically broadcasts :)